Guide/file-storage.markdown

```diff
@@ -575,3 +575,244 @@ config = do
+
+                renderFile (cs $ storagePrefix <> storedFile.path) "application/jpg"
+
+-- | Extracts the directory and UUID from a path like "pictures/8ed22caa-11ea-4c45-a05e-91a51e72558d"
+extractDirectoryAndUUID :: (?context :: context, ConfigProvider context) => Text -> (Text, Text)
+extractDirectoryAndUUID inputText =
+    case reverse parts of
+        uuid : pathSegments -> (T.intercalate "/" (reverse pathSegments), uuid)
+        _ -> ("", "")
+    where
+        frameworkConfig = ?context.frameworkConfig
+        trimmedText = T.replace (frameworkConfig.baseUrl <> "/") "" inputText
+        parts = T.splitOn "/" trimmedText
+```
+
+Now, from any `Show` action, we can use the image style. Here we create a 400px x 200px image style for the original image.
+
+```haskell
+[hsx|<img src={pathTo $ RenderImageStyleAction 400 200 imageUrl} />|]
+where
+    imageUrl = "http://localhost:8000/static/picture/b4c8f55c-16d6-41f0-9503-77352b134e14"
+```
+
+### Security Considerations
+
+The above implementation is not secure. Anyone can request any image style of any image. This may put a burden on the site. Imagine a malicious user requesting thousands of different dimensions of a 100MB image. The server would have to generate thousands of images. This would put a lot of load on the server. To prevent this, we need to add some access control. The way we would do it, as that is the simplest way, is to add a signed token to the image style URL. The token would be generated by the server and would contain the original image path, the desired width and height. The server would then verify the token and only serve the image style if the token is valid. This way, the server would not have to generate the image style on the fly, but only if the token is valid. If the token is not valid, then the server would return an access denied.
+
+We start by adding the `cryptonite` package to the `default.nix` file, and re-start `./start` to get the package.
+
+```nix
+# default.nix
+haskellEnv = import "${ihp}/NixSupport/default.nix" {
+    ihp = ihp;
+    haskellDeps = p: with p; [
+        cabal-install
+        base
+        # ...
+        cryptonite # <--- ADD THIS
+    ];
+```
+
+Then we need to generate a private and public key pair. We can do it in our Config file, and save the keys for later use.
+
+```haskell
+-- Config/Config.hs
+import Crypto.PubKey.RSA as RSA
+
+newtype RsaPublicAndPrivateKeys = RsaPublicAndPrivateKeys (RSA.PublicKey, RSA.PrivateKey)
+
+config :: ConfigBuilder
+config = do
+    -- See https://ihp.digitallyinduced.com/Guide/config.html
+    -- for what you can do here
+
+    -- ...
+
+    -- Private and public keys to sign and verify image style URLs.
+    -- The 300 means how large your URL can be.
+    -- The 65537 is the public exponent of RSA.
+    -- You can probably leave those values as is.
+    (publicKey, privateKey) <- liftIO $ liftIO $ RSA.generate 300 65537
```

Comment:
This generates a new RSA key every time the app restarts. This invalidates all existing image urls generated, which could affect the end user experience in bad ways. It would be better to generate the RSA key only once and store it in `Config/image_signing_key.rsa` and then load it from here. The key can be generated by using `openssl` from the CLI, or even by some Haskell code here inside the `Config.hs` that checks whether the file already exists.

Reply:
:+1: I'll try to store it under `Config/private_key.rsa` (as it could be used for other stuff not just image signing).

Also, I think it's nicer that we do it on Config, rather than asking user to `openssl` -- so one could only copy/ paste code - and it should work.

Reply:
I've used the JWT package in the past and it also uses RSA keys for signing. And it also uses cryptonite, so maybe we can just copy how the JWT package is doing it: https://github.com/puffnfresh/haskell-jwt/blob/master/src/Web/JWT.hs

Especially the [`readRsaSecret`](https://github.com/puffnfresh/haskell-jwt/blob/master/src/Web/JWT.hs#L436) function is basically doing what we want.

Reply:
Got it working, thanks